File: runtime/python/units/wasi/fd_renumber.py
def wasi_fd_renumber(self, fd, to):
    # Both endpoints must be open fds: renumbering onto an unused number is BADF.
    if fd not in self.fds or to not in self.fds:
        return self.ERRNO_BADF
    dst = self.fds[to]
    if not isinstance(dst, self.WasiDir) and dst not in self.std_ios:
        dst.close()
    self.fds[to] = self.fds[fd]
    self.fd_meta[to] = self.fd_meta[fd]
    del self.fds[fd]
    del self.fd_meta[fd]
    return self.ERRNO_SUCCESS

File: runtime/perl/units/wasi/fd_renumber.pl
# requires:
sub wasi_fd_renumber {
    my ($self, $fd, $to) = @_;
    # Both endpoints must be open fds: renumbering onto an unused number is BADF.
    return ERRNO_BADF unless exists $self->{fds}{$fd} && exists $self->{fds}{$to};
    my $dst = $self->{fds}{$to};
    close($dst->{fh}) if !$dst->{dir} && !defined($dst->{std});
    $self->{fds}{$to} = delete $self->{fds}{$fd};
    $self->{meta}{$to} = delete $self->{meta}{$fd};
    return ERRNO_SUCCESS;
}
